components/psst/browser/core/psst_rule_registry_unittest.cc

```diff
@@ -0,0 +1,166 @@
+  if (!success || contents.empty()) {
+    VLOG(2) << "ReadFile: cannot " << "read file " << file_path;
+  }
+  return contents;
+}
+}  // namespace
+
+class MockRuleDataReader : public RuleDataReader {
+ public:
+  explicit MockRuleDataReader(const base::FilePath& component_path);
+  ~MockRuleDataReader() override = default;
+
+  MOCK_METHOD(std::optional<std::string>,
+              ReadUserScript,
+              (const PsstRule& rule),
+              (override, const));
+  MOCK_METHOD(std::optional<std::string>,
+              ReadPolicyScript,
+              (const PsstRule& rule),
+              (override, const));
+};
+MockRuleDataReader::MockRuleDataReader(const base::FilePath& component_path)
+    : RuleDataReader(component_path) {}
+
+class MockPsstRuleRegistryImpl : public PsstRuleRegistryImpl {
+ public:
+  MOCK_METHOD(void, OnLoadRules, (const std::string& data), (override));
+};
+
+using OnRuleMatchedCallback = base::RepeatingCallback<void(
+    const std::optional<MatchedRule>& matched_rule)>;
+
+class PsstRuleRegistryUnitTest : public testing::Test {
+ public:
+  void SetUp() override {
+    scoped_feature_list_.InitAndEnableFeature(features::kBravePsst);
+
+    base::FilePath test_data_dir(
+        base::PathService::CheckedGet(brave::DIR_TEST_DATA));
+    test_data_dir_base_ = test_data_dir.AppendASCII("psst-component-data");
+
+    auto rule_data_reader_mock = std::make_unique<MockRuleDataReader>(
+        base::FilePath(test_data_dir_base_));
+    rule_data_reader_mock_ = rule_data_reader_mock.get();
+
+    auto psst_rul_registry = std::make_unique<MockPsstRuleRegistryImpl>();
+    psst_rul_registry_ = psst_rul_registry.get();
+
+    PsstRuleRegistryAccessor::GetInstance()->SetRegistryForTesting(
+        std::move(psst_rul_registry));
+
+    psst_rul_registry_->SetRuleDataReaderForTest(
+        std::move(rule_data_reader_mock));
+  }
+
+  MockRuleDataReader* GetMockRuleDataReader() { return rule_data_reader_mock_; }
+
+  base::FilePath GetTestDataDirBase() const { return test_data_dir_base_; }
+
+  MockPsstRuleRegistryImpl* GetPsstRuleRegistry() { return psst_rul_registry_; }
+
+  void MockOnLoadRules(base::RunLoop& run_loop) {
+    EXPECT_CALL(*GetPsstRuleRegistry(), OnLoadRules(testing::_))
+        .WillOnce(testing::Invoke([&](const std::string& data) {
+          EXPECT_EQ(data, ReadFile(GetTestDataDirBase().Append(
+                              base::FilePath::FromUTF8Unsafe("psst.json"))));
+          (*GetPsstRuleRegistry()).PsstRuleRegistryImpl::OnLoadRules(data);
+          run_loop.Quit();
+        }));
+  }
+
+  void TestMatchedRuleLoading(const GURL& url,
+                              std::optional<std::string> user_script,
+                              std::optional<std::string> policy_script,
+                              OnRuleMatchedCallback on_rule_matched_callback) {
+    {
+      base::RunLoop run_loop;
+      EXPECT_CALL(*GetMockRuleDataReader(), ReadUserScript)
```

Comment:
please don't put expectations inside a helper method because it obscures the location of the actual failure